File: lap-reference/src/crypto-util.js
import { createHash, createPublicKey, verify as cryptoVerify } from "node:crypto";

const B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
// DER SPKI header for a raw Ed25519 public key (RFC 8410)
const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export function b64urlDecode(s) {
  return Buffer.from(s, "base64url");
}

export function b64urlEncode(buf) {
  return Buffer.from(buf).toString("base64url");
}

export function sha256Hex(data) {
  return createHash("sha256").update(data).digest("hex");
}

export function sha256B64url(data) {
  return b64urlEncode(createHash("sha256").update(data).digest());
}

export function b58Encode(buf) {
  let n = BigInt("0x" + (Buffer.from(buf).toString("hex") || "0"));
  let s = "";
  while (n > 0n) {
    const r = n % 58n;
    n /= 58n;
    s = B58[Number(r)] + s;
  }
  let pad = 0;
  for (const byte of buf) {
    if (byte === 0) pad++;
    else break;
  }
  return "1".repeat(pad) + s;
}

export function b58Decode(s) {
  if (!s || s.length === 0) return Buffer.alloc(0);
  let n = 0n;
  for (const ch of s) {
    const v = B58.indexOf(ch);
    if (v < 0) throw new Error(`invalid base58 character: ${ch}`);
    n = n * 58n + BigInt(v);
  }
  let pad = 0;
  for (const ch of s) {
    if (ch === "1") pad++;
    else break;
  }
  if (n === 0n) return Buffer.alloc(pad);
  let hex = n.toString(16);
  if (hex.length % 2) hex = "0" + hex;
  const out = Buffer.from(hex, "hex");
  return pad ? Buffer.concat([Buffer.alloc(pad), out]) : out;
}

// did:key:z6Mk... -> 32-byte raw Ed25519 public key
export function didKeyToRawPublicKey(did) {
  if (!did.startsWith("did:key:z")) throw new Error(`unsupported DID method: ${did}`);
  const multibase = did.slice("did:key:z".length);
  // F12: bound input BEFORE base58 decoding — an ed25519 did:key multibase is ~46-48
  // chars; `iss`/`agentDid` are attacker-controlled and decoded pre-auth, so an
  // unbounded string would drive O(n^2) BigInt work (CPU-exhaustion DoS).
  if (multibase.length > 48) throw new Error("LAP_ERR_DID: multibase too long for ed25519 did:key");
  const decoded = b58Decode(multibase);
  if (decoded[0] !== 0xed || decoded[1] !== 0x01) throw new Error("not an ed25519-pub did:key");
  const raw = decoded.subarray(2);
  if (raw.length !== 32) throw new Error(`bad key length: ${raw.length}`);
  return raw;
}

export function publicKeyFromRaw(raw32) {
  return createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, raw32]),
    format: "der",
    type: "spki",
  });
}

export function publicKeyFromDid(did) {
  return publicKeyFromRaw(didKeyToRawPublicKey(did));
}

// 32-byte raw Ed25519 public key -> did:key:z6Mk...
export function didKeyFromRawPublicKey(raw32) {
  return "did:key:z" + b58Encode(Buffer.concat([Buffer.from([0xed, 0x01]), raw32]));
}

export function rawPublicKeyFromKeyObject(publicKey) {
  const der = publicKey.export({ format: "der", type: "spki" });
  return der.subarray(der.length - 32);
}

export function verifyEd25519(publicKey, data, signature) {
  return cryptoVerify(null, Buffer.from(data), publicKey, signature);
}

// DID-safe normalization: NFC + lowercase the "did:method:" prefix ONLY.
// The method-specific id is case-sensitive (base58 in did:key) and MUST NOT be
// case-folded — lowercasing it would corrupt the key. (Round-3 triage note: the
// reviewer's proposed fix of running DIDs through normalizeUri would do exactly that.)
export function normalizeDid(did) {
  const s = did.normalize("NFC");
  const m = s.match(/^did:([A-Za-z0-9]+):(.+)$/);
  if (!m) return s;
  return "did:" + m[1].toLowerCase() + ":" + m[2];
}

// LIP-1 §3 string normalization for resource URIs: NFC + lowercase scheme/host.
// Do NOT use on DIDs — use normalizeDid (host-lowercasing corrupts did:key ids).
export function normalizeUri(uri) {
  const s = uri.normalize("NFC");
  const m = s.match(/^([a-zA-Z][a-zA-Z0-9+.-]*):(\/\/)?([^/]*)(.*)$/);
  if (!m) return s;
  const [, scheme, slashes = "", host, rest] = m;
  return scheme.toLowerCase() + ":" + slashes + host.toLowerCase() + rest;
}

// JCS-compatible canonical JSON for the LAP data subset (ASCII strings, integers,
// objects, arrays, null, booleans). Matches Python json.dumps(sort_keys=True,
// separators=(",",":")) for this subset.
export function jcs(value) {
  if (value === null || typeof value === "boolean") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isInteger(value)) throw new Error("JCS subset: integers only");
    return String(value);
  }
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) return "[" + value.map(jcs).join(",") + "]";
  if (typeof value === "object") {
    const keys = Object.keys(value).sort();
    return "{" + keys.map((k) => JSON.stringify(k) + ":" + jcs(value[k])).join(",") + "}";
  }
  throw new Error(`JCS subset: unsupported type ${typeof value}`);
}

File: lap-python/living_agents/crypto_util.py
"""
LIP-1 / LIP-4 Cryptographic utilities: Base58, JCS canonicalization, Ed25519 did:key parsing.
"""

import base64
import hashlib
import json
import re
from typing import Any, Union

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

B58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


_B64URL_ALPHABET = re.compile(r"^[A-Za-z0-9_-]*$")


def b64url_decode(s: str) -> bytes:
    """Decode unpadded base64url STRICTLY (RFC 4648 §5; §3.5 canonical trailing bits).

    Lenient decoding lets one signature have many textual forms — 'A'->'B' as the final
    character of an Ed25519 signature changes padding bits only and still verifies — so
    anything keyed by the TEXT (passport hashes, dedup, denylists, transparency-log entries)
    could be evaded by re-encoding. Reject padding, foreign characters, and non-zero
    trailing bits instead; the check is "re-encode and compare".
    """
    if not isinstance(s, str) or not _B64URL_ALPHABET.match(s) or len(s) % 4 == 1:
        raise ValueError("LAP_ERR_ENCODING: invalid base64url")
    raw = base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))
    if b64url_encode(raw) != s:
        raise ValueError("LAP_ERR_ENCODING: non-canonical base64url")
    return raw

def b64url_encode(buf: bytes) -> str:
    """Encode bytes to unpadded base64url string."""
    return base64.urlsafe_b64encode(buf).decode("ascii").rstrip("=")


def sha256_hex(data: Union[bytes, str]) -> str:
    """Return hex-encoded SHA-256 digest."""
    buf = data.encode("utf-8") if isinstance(data, str) else data
    return hashlib.sha256(buf).hexdigest()


def sha256_b64url(data: Union[bytes, str]) -> str:
    """Return base64url-encoded SHA-256 digest."""
    buf = data.encode("utf-8") if isinstance(data, str) else data
    return b64url_encode(hashlib.sha256(buf).digest())


def b58encode(buf: bytes) -> str:
    """Encode bytes to Bitcoin Base58 string with zero-byte preservation."""
    if not buf:
        return ""
    n = int.from_bytes(buf, byteorder="big")
    chars = []
    while n > 0:
        n, r = divmod(n, 58)
        chars.append(B58_ALPHABET[r])
    chars.reverse()
    
    pad = 0
    for byte in buf:
        if byte == 0:
            pad += 1
        else:
            break
    return ("1" * pad) + "".join(chars)


def b58decode(s: str) -> bytes:
    """Decode Bitcoin Base58 string to bytes with zero-byte preservation."""
    if not s:
        return b""
    n = 0
    for ch in s:
        v = B58_ALPHABET.find(ch)
        if v < 0:
            raise ValueError(f"invalid base58 character: {ch}")
        n = n * 58 + v

    pad = 0
    for ch in s:
        if ch == "1":
            pad += 1
        else:
            break

    if n == 0:
        return b"\x00" * pad

    hex_str = hex(n)[2:]
    if len(hex_str) % 2:
        hex_str = "0" + hex_str
    out = bytes.fromhex(hex_str)
    return (b"\x00" * pad) + out if pad else out


def did_key_to_raw_public_key(did: str) -> bytes:
    """Parse did:key:z6Mk... to 32-byte raw Ed25519 public key."""
    if not did.startswith("did:key:z"):
        raise ValueError(f"unsupported DID method: {did}")
    multibase = did[len("did:key:z"):]
    # F12: bound input BEFORE base58 decoding — an ed25519 did:key multibase is ~46-48
    # chars; iss/agent_did are attacker-controlled and decoded pre-auth, so an unbounded
    # string would drive O(n^2) big-int work (CPU-exhaustion DoS).
    if len(multibase) > 48:
        raise ValueError("LAP_ERR_DID: multibase too long for ed25519 did:key")
    decoded = b58decode(multibase)
    if len(decoded) < 2 or decoded[0] != 0xED or decoded[1] != 0x01:
        raise ValueError("not an ed25519-pub did:key (expected multicodec prefix 0xed, 0x01)")
    raw = decoded[2:]
    if len(raw) != 32:
        raise ValueError(f"bad key length: {len(raw)} (expected 32 bytes)")
    return raw


def public_key_from_raw(raw32: bytes) -> Ed25519PublicKey:
    """Create Ed25519PublicKey from raw 32 bytes."""
    return Ed25519PublicKey.from_public_bytes(raw32)


def public_key_from_did(did: str) -> Ed25519PublicKey:
    """Resolve did:key to Ed25519PublicKey."""
    return public_key_from_raw(did_key_to_raw_public_key(did))


def did_key_from_raw_public_key(raw32: bytes) -> str:
    """Convert raw 32-byte Ed25519 public key to did:key:z6Mk..."""
    multicodec = b"\xed\x01" + raw32
    return "did:key:z" + b58encode(multicodec)


def verify_ed25519(public_key: Ed25519PublicKey, data: bytes, signature: bytes) -> bool:
    """Verify Ed25519 signature over data bytes."""
    try:
        public_key.verify(signature, data)
        return True
    except InvalidSignature:
        return False


def normalize_did(did: str) -> str:
    """
    DID-safe normalization: NFC + lowercase the 'did:method:' prefix ONLY.
    Preserves case of the method-specific identifier.
    """
    import unicodedata
    s = unicodedata.normalize("NFC", did)
    m = re.match(r"^did:([A-Za-z0-9]+):(.+)$", s, re.IGNORECASE)
    if not m:
        return s
    return f"did:{m.group(1).lower()}:{m.group(2)}"


def normalize_uri(uri: str) -> str:
    """
    LIP-1 §3 string normalization for resource URIs: NFC + lowercase scheme and host.
    """
    import unicodedata
    s = unicodedata.normalize("NFC", uri)
    m = re.match(r"^([a-zA-Z][a-zA-Z0-9+.-]*):(//)?([^/]*)(.*)$", s)
    if not m:
        return s
    scheme, slashes, host, rest = m.group(1), m.group(2) or "", m.group(3), m.group(4)
    return f"{scheme.lower()}:{slashes}{host.lower()}{rest}"


def jcs(value: Any) -> str:
    """
    JCS (RFC 8785) canonical JSON serialization for the LAP data subset.
    Produces byte-identical output to Node.js reference implementation.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if isinstance(value, float):
        if not value.is_integer():
            raise ValueError("JCS subset: integers only")
        return str(int(value))
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(jcs(x) for x in value) + "]"
    if isinstance(value, dict):
        sorted_keys = sorted(value.keys())
        items = [json.dumps(k, ensure_ascii=False) + ":" + jcs(value[k]) for k in sorted_keys]
        return "{" + ",".join(items) + "}"
    raise TypeError(f"JCS subset: unsupported type {type(value)}")
